Sources/Foundation/JSONSerialization.swift

```diff
@@ -549,523 +621,838 @@ private struct JSONWriter {
-
-    static let whitespaceASCII: [UInt8] = [
-        0x09, // Horizontal tab
-        0x0A, // Line feed or New line
-        0x0D, // Carriage return
-        0x20, // Space
-    ]
-
-    struct Structure {
-        static let BeginArray: UInt8     = 0x5B // [
-        static let EndArray: UInt8       = 0x5D // ]
-        static let BeginObject: UInt8    = 0x7B // {
-        static let EndObject: UInt8      = 0x7D // }
-        static let NameSeparator: UInt8  = 0x3A // :
-        static let ValueSeparator: UInt8 = 0x2C // ,
-        static let QuotationMark: UInt8  = 0x22 // "
-        static let Escape: UInt8         = 0x5C // \
+private struct JSONParser {
+    
+    var reader: DocumentReader
+    var depth: Int = 0
+    
+    init<Bytes: Collection>(bytes: Bytes) where Bytes.Element == UInt8 {
+        self.reader = DocumentReader(bytes: [UInt8](bytes))
     }
 
-    typealias Index = Int
-    typealias IndexDistance = Int
-
-    struct UnicodeSource {
-        let buffer: UnsafeBufferPointer<UInt8>
-        let encoding: String.Encoding
-        let step: Int
-
-        init(buffer: UnsafeBufferPointer<UInt8>, encoding: String.Encoding) {
-            self.buffer = buffer
-            self.encoding = encoding
-
-            self.step = {
-                switch encoding {
-                case .utf8:
-                    return 1
-                case .utf16BigEndian, .utf16LittleEndian:
-                    return 2
-                case .utf32BigEndian, .utf32LittleEndian:
-                    return 4
-                default:
-                    return 1
-                }
-            }()
+    mutating func parse() throws -> JSONValue {
+        let value = try parseValue()
+        #if DEBUG
+        defer {
+            guard self.depth == 0 else {
+                preconditionFailure("Expected to end parsing with a depth of 0")
+            }
         }
+        #endif
 
-        func takeASCII(_ input: Index) -> (UInt8, Index)? {
-            guard hasNext(input) else {
-                return nil
+        // handle extra character if top level was number
+        if case .number = value {
+            guard let extraCharacter = reader.value else {
+                return value
             }
 
-            let index: Int
-            switch encoding {
-            case .utf8:
-                index = input
-            case .utf16BigEndian where buffer[input] == 0:
-                index = input + 1
-            case .utf32BigEndian where buffer[input] == 0 && buffer[input+1] == 0 && buffer[input+2] == 0:
-                index = input + 3
-            case .utf16LittleEndian where buffer[input+1] == 0:
-                index = input
-            case .utf32LittleEndian where buffer[input+1] == 0 && buffer[input+2] == 0 && buffer[input+3] == 0:
-                index = input
+            switch extraCharacter {
+            case UInt8(ascii: " "), UInt8(ascii: "\r"), UInt8(ascii: "\n"), UInt8(ascii: "\t"):
```

Comment:
These should be defined as constants to make it easier to read and reduce errors due to a typo with a characters. Eventually we could merge and reuse the constants across other parsers in sclf.